client/src/components/FeedbackPage/FeedbackFormSection/FeedbackFormSection.tsx

```diff
@@ -53,63 +75,209 @@ const feedbackFormConfig: FeedbackForm = {
+      ...feedbackForm,
+      [inputName]: {
+        ...feedbackForm[inputName],
+        value: inputValue,
+        valid: isValid,
+        errMsg: message
+      }
+    });
+  };
+
+  const createUpdatedForm = (
+    validatedInputs: { valid: boolean; errMsg: string }[],
+    inputNames: FeedbackFormInput[],
+    form: FeedbackForm
+  ) =>
+    validatedInputs.reduce(
+      (prevFields, currField, currentIndex) => ({
+        ...prevFields,
+        [inputNames[currentIndex]]: {
+          ...form[inputNames[currentIndex]],
+          ...currField
+        }
+      }),
+      {}
+    );
+
+  const handleFormChange = () => {
+    const inputNames = Object.keys(FeedbackFormInput).map(key => key as FeedbackFormInput);
+
+    const validatedInputs = feedbackFormInputs.map((input, i) => {
+      const { isValid: valid, message: errMsg } = inputValidator(
+        input.value!,
+        inputNames[i],
+        feedbackFormConfig[inputNames[i]].validation
+      );
+      return { valid, errMsg };
+    });
+
+    const updatedForm = createUpdatedForm(validatedInputs, inputNames, feedbackForm);
+
+    setFeedbackForm({
+      ...feedbackForm,
+      ...updatedForm
+    });
+  };
+
+  const handleSubmitForm = (e: FormEvent<HTMLFormElement>) => {
+    e.preventDefault();
+
+    handleFormChange();
+
+    if (isFormValid) {
+      // I will write code here in another feature
+    }
+  };
+
+  useEffect(() => {
+    setIsFormDirty(feedbackFormInputs.some(input => input.value));
+  }, [feedbackFormInputs]);
+
+  useEffect(() => {
+    setIsFormValid(!feedbackFormInputs.map(input => input.valid).includes(false));
+  }, [feedbackFormInputs]);
+
+  useBeforeUnload({ when: isFormDirty });
 
   return (
     <S.FeedbackFormSection id="survey">
       <Container>
         <Flex directionColumn>
           <h2>Fill Form Below</h2>
-          <form>
-            {formElementsArray.map(({ id, config }) => (
-              <S.FeedbackFormGroup key={id}>
-                <S.FeedbackFormLabel htmlFor={id}>{config.label}</S.FeedbackFormLabel>
-                <Input inputType={config.inputType} id={id} inputConfig={config.inputConfig} />
-              </S.FeedbackFormGroup>
-            ))}
+          <form onSubmit={handleSubmitForm}>
+            <Flex directionColumn alignFlexStart>
+              {formElementsArray.map(({ id, config }, i) => (
+                <S.FeedbackFormGroup key={id}>
+                  <S.FeedbackFormLabel htmlFor={id}>{`${i + 1}. ${config.label} ${
+                    config.validation?.required ? '' : '(optional)'
+                  }`}</S.FeedbackFormLabel>
+                  <Input
+                    inputType={config.inputType}
+                    id={id}
+                    name={id}
+                    inputConfig={config.inputConfig}
+                    onChange={handleInputChange}
+                  />
+                  <S.FeedbackFormError>
+                    {feedbackForm[id as FeedbackFormInput].errMsg}
```

Comment:
I don't know how. If I leave it as `feedbackForm[id]`, I get compilation error "Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'FeedbackForm'. No index signature with a parameter of type 'string' was found on type 'FeedbackForm'." But I need this to be dynamic.

What about `{feedbackForm[FeedbackFormInput[id as keyof typeof FeedbackFormInput]].errMsg}`? (https://stackoverflow.com/questions/50417254/dynamically-access-enum-in-typescript-by-key)

Reply:
please provide me exact code line/reason, why typescript thinks that `id` is a string, compilation error is not a reason to add `as` everywhere. yes, sometimes it's the only solution, but first you need to know why it was happened 

`{feedbackForm[FeedbackFormInput[id as keyof typeof FeedbackFormInput]].errMsg}` doesn't make code better

Reply:
```
const formElementsArray = Object.entries(feedbackFormConfig).map(entry => ({
  id: entry[0],
  config: entry[1]
}));
````
The above code makes typescript think `id` is a string, because `Object.entries()` converts enum keys into strings, so does `Object.keys()`. And in this case, `entry[0]` is a key of enum `FeedbackFormInput`. I can't seem to find any other workaround that does not involve using keyword `as`.

https://stackoverflow.com/questions/55012174/why-doesnt-object-keys-return-a-keyof-type-in-typescript
https://dev.to/prod/mastering-enums-in-typescript-1c1j